src/client/java/teammates/client/scripts/sql/DataMigrationForCourseEntitySql.java

```diff
@@ -148,4 +186,200 @@ private Student createStudent(teammates.storage.sqlentity.Course newCourse,
+                } else {
+                    deleteCourseCascade(currentCourse);
+                }
+            }
+
+            while (currentCourse != null) {
+                shouldContinue = true;
+                migrateWithoutTrx(currentCourse);
+                numberOfScannedKey.incrementAndGet();
+
+                cursor = iterator.getCursorAfter();
+                savePositionOfCursorToFile(cursor);
+
+                currentCourse = iterator.hasNext() ? (Course) iterator.next() : null;
+            }
+
+            if (shouldContinue) {
+                log(String.format("Cursor Position: %s", cursor.toUrlSafe()));
+                log(String.format("Number Of Course Entity Key Scanned: %d", numberOfScannedKey.get()));
+                log(String.format("Number Of Course Entity affected: %d", numberOfAffectedEntities.get()));
+                log(String.format("Number Of Course Entity updated: %d", numberOfUpdatedEntities.get()));
+            }
+        }
+
+        deleteCursorPositionFile();
+        log(isPreview() ? "Preview Completed!" : "Migration Completed!");
+        log("Total number of course entities: " + numberOfScannedKey.get());
+        log("Number of affected course entities: " + numberOfAffectedEntities.get());
+        log("Number of updated course entities: " + numberOfUpdatedEntities.get());
+    }
+
+    /**
+     * Deletes the course and its related entities from sql database.
+     */
+    private void deleteCourseCascade(Course oldCourse) {
+        // TODO: Implement deleteCourseCascase
+        log("delete course id: " + oldCourse.getUniqueId());
+    }
+
+    /**
+     * Reads the cursor position from the saved file.
+     *
+     * @return cursor if the file can be properly decoded.
+     */
+    private Optional<Cursor> readPositionOfCursorFromFile() {
+        try {
+            String cursorPosition = FileHelper.readFile(BASE_LOG_URI + this.getClass().getSimpleName() + ".cursor");
+            return Optional.of(Cursor.fromUrlSafe(cursorPosition));
+        } catch (IOException | IllegalArgumentException e) {
+            return Optional.empty();
+        }
+    }
+
+    /**
+     * Saves the cursor position to a file so it can be used in the next run.
+     */
+    private void savePositionOfCursorToFile(Cursor cursor) {
+        try {
+            FileHelper.saveFile(
+                    BASE_LOG_URI + this.getClass().getSimpleName() + ".cursor", cursor.toUrlSafe());
+        } catch (IOException e) {
+            logError("Fail to save cursor position " + e.getMessage());
+        }
+    }
+
+    /**
+     * Deletes the cursor position file.
+     */
+    private void deleteCursorPositionFile() {
+        FileHelper.deleteFile(BASE_LOG_URI + this.getClass().getSimpleName() + ".cursor");
+    }
+
+    /**
+     * Migrates the entity without transaction for better performance.
+     */
+    private void migrateWithoutTrx(Course entity) {
+        doMigration(entity);
+    }
```

Comment:
Probably can remove this if we're not using it for this migration, and just call `doMigration` directly